### PR TITLE
Make sure 'in' is honored in cref parsers/completion lists

### LINKED ISSUE #23957: XML code comments and 'in' parameters
@andi6 commented on [Thu Dec 21 2017](https://github.com/dotnet/csharplang/issues/1209)

I don't find a way to reference a method with an 'in' parameter in xml code comments. Also VisualStudio (15.5.2) seems to be unable to do it.

If I take a sample method:

`public void MethodWithInPar(in double dummy) { }`

VisualStudio suggests

```
		/// <summary>
		/// This is the summary for <see cref="MethodWithInPar(double)"/>.
		/// </summary>
```

with xml comment generation enabled in project settings (classic .NET framework project) I get CS1574.

---

@HaloFour commented on [Thu Dec 21 2017](https://github.com/dotnet/csharplang/issues/1209#issuecomment-353348770)

@andi6 

What about:

```cs
// <summary>
/// This is the summary for <see cref="MethodWithInPar(ref double)"/>.
/// </summary>
```

---

@andi6 commented on [Thu Dec 21 2017](https://github.com/dotnet/csharplang/issues/1209#issuecomment-353368592)

Sorry, wrong button

@HaloFour 

I've tried this also: CS1574

### PR DESCRIPTION
**Pull request by OmarTawfik**
### Customer scenario

`RefKind.In` is not honored in cref comments. This leads to:
1) methods with `in` parameters are populated as regular byval parameters.
2) compiler displays warnings when user manually writes `<see cref="M(in int x)" />`

Note: API change will follow #22472

### Bugs this fixes

Fixes #23957 

### Workarounds, if any

None.

### Risk

Minimal. It is an addition to the parser/completion lists.

### Performance impact

Minimal. No complexity changes.

### Is this a regression from a previous update?

No.

### Root cause analysis

New feature. New tests were added to guard against possible future regressions.

### How was the bug found?

Customer comment on GitHub.

### Test documentation updated?

----

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@gafter for compiler API change. I tried following #22472
@dotnet/roslyn-compiler for small compiler change
@dotnet/roslyn-ide for IDE change. Not sure if there are other APIs I should be checking/adding tests for.

**Comment by OmarTawfik** [comment]
Also, for API check tool, shouldn't it detect the new public methods I added (extensions)?

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,6 +1,10 @@
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax.Update(Microsoft.CodeAnalysis.CSharp.Syntax.NameColonSyntax nameColon, Microsoft.CodeAnalysis.SyntaxToken refOrOutKeyword, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression) -> Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax.WithRefKindKeyword(Microsoft.CodeAnalysis.SyntaxToken refOrOutKeyword) -> Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax
+*REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.CrefParameterSyntax.RefOrOutKeyword.get -> Microsoft.CodeAnalysis.SyntaxToken
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: Has this already shipped?  If so, isn't this a breaking change?

File: src/Compilers/CSharp/Portable/Syntax/CrefParameterSyntax.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Microsoft.CodeAnalysis.CSharp.Syntax
+{
+    public sealed partial class CrefParameterSyntax
+    {
+        /// <summary>
+        /// Pre C# 7.2 back-compat overload, which simply calls the replacement property <see cref="RefKindKeyword"/>.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public SyntaxToken RefOrOutKeyword
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/CrefParameterSyntax.cs]: i'm confused.  the API checker said this was removed...

File: src/Compilers/CSharp/Portable/Syntax/CrefParameterSyntax.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            get => this.RefKindKeyword;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/CrefParameterSyntax.cs]: nit, just use => for the property and elide the accessor.

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: @jasonmalinowski did we update the API files when we shipped 15.5?

**Comment by CyrusNajmabadi** [comment]
IDE side LGTM.

**Comment by agocke** [comment]
My expectation for the API is we keep the old one and add the new one. It looks like that's what the code does, so I'm not sure what's the deal with the public API checker. LGTM once that's figured out.

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: Will keep the old APIs. Looking into why the tool is behaving this way.

---
In reply to: [159702342](https://github.com/dotnet/roslyn/pull/24023#discussion_r159702342) [](ancestors = 159702342)

**Comment by OmarTawfik** [comment]
Fixed the API files. I discovered a bug in the API tool, as it should report errors when items are marked as *removed* while they still exist in source code. Where do we report such bugs?

**Comment by jaredpar** [comment]
@OmarTawfik report bugs on the analyzer at https://github.com/dotnet/roslyn-analyzers. This is also where the source code for it lives.

File: src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs [hunk 4]
```diff
@@ -1156,20 +1156,18 @@ private BaseCrefParameterListSyntax ParseBaseCrefParameterList(bool useSquareBra
         /// <summary>
         /// True if the current token could be the beginning of a cref parameter.
         /// </summary>
-        private bool IsPossibleCrefParameter
+        private bool IsPossibleCrefParameter()
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs]: Why did this change to a method?

**Comment by OmarTawfik** [comment]
Filed https://github.com/dotnet/roslyn-analyzers/issues/1506 for the analyzer.

File: src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs (hunk 4 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs]: Aligns with our code style in the rest of the parser. No reason to keep it as a property.

---
In reply to: [159807126](https://github.com/dotnet/roslyn/pull/24023#discussion_r159807126) [](ancestors = 159807126)

**Comment by OmarTawfik** [comment]
@jasonmalinowski for IDE sign off.
@gafter for API sign off.

**Comment by OmarTawfik** [comment]
cc @dotnet/roslyn-compat-council as well

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-ide this is ready for review
@gafter for API change

**Comment by Pilchie** [comment]
I'm okay with the breaking change given that it is source only when named parameters are used.

However, you'll need ask/escrow approval, and to figure out what branch to merge into.

**Comment by jcouv** [comment]
@OmarTawfik If you're targeting 15.6, don't forget to retarget the PR (to `dev15.6.x` branch).

**Comment by OmarTawfik** [comment]
@Pilchie I believe 15.6 is still ask mode, and we will enter ESCROW Wednesday PM. Isn't that still the case? cc @jasonmalinowski 

If so, @MeiChin-Tsai and @Pilchie, this is ready for ask-mode approval.

File: docs/compilers/CSharp/Compiler Breaking Changes - post VS2017.md [hunk 5]
```diff
@@ -77,3 +77,5 @@ In Visual Studio 2017 version 15.6 such conversions will be explicitly disallowe
 Example: `Func<int> f = default(TypedReference).GetHashCode; // new error CS0123: No overload for 'GetHashCode' matches delegate 'Func<int>'` 
    
 - https://github.com/dotnet/roslyn/pull/23416 Before Visual Studio 2017 version 15.6 (Roslyn version 2.8) the compiler accepted `__arglist(...)` expressions with void-typed arguments. For instance, `__arglist(Console.WriteLine())`. But such program would fail at runtime. In Visual Studio 2017 version 15.6, this causes a compile-time error.
+
+- https://github.com/dotnet/roslyn/pull/24023 Before Visual Studio 2017 version 15.6, Microsoft.CodeAnalysis.CSharp.Syntax.CrefParameterSyntax constructor and Update(), the parameter refOrOutKeyword was renamed to refKindKeyword (source breaking change if you're using named arguments).
```
Comment (jcouv) [inline docs/compilers/CSharp/Compiler Breaking Changes - post VS2017.md]: Probably should be "In Visual Studio ..."

**Comment by Pilchie** [comment]
Approved for 15.6 Preview 4.

**Pull request merged**